numbat/src/parser.rs

```diff
@@ -2,29 +2,56 @@
 //!
 //! Grammar:
 //! ```txt
-//! statement       →   expression | variable_decl | function_decl | dimension_decl | unit_decl | procedure_call | module_import
+//! statement       →   variable_decl | function_decl | dimension_decl | decorator | unit_decl | module_import | procedure_call | expression
+//!
+//! variable_decl   →   "let" identifier ( ":" type_annotation ) ? "=" expression
+//! type_annotation →   boolean | string | dimension_expr
+//! function_decl   →   "fn" identifier ( fn_decl_generic ) ? fn_decl_param "->" type_annotation "=" expression
+//! fn_decl_generic →   "<" ( identifier "," ) * identifier ? ">"
+//! fn_decl_param   →   "(" ( identifier ( ":" dimension_expr ) ? "," )* ( identifier ( ":" dimension_expr ) ) ? ")"
+//! dimension_decl  →   "dimension" identifier ( "=" dimension_expr ) *
+//! decorator       →   "@" ( "metric_prefixes" | "binary_prefixes" | ( "aliases(" list_of_alsiases ")" ) )
+//! unit_decl       →   "unit" ( ":" dimension_expr ) ? ( "=" expression ) ?
+//! procedure_call  →   ( "print" | "assert_eq" | "type" ) ( "(" arguments? ")" ) ?
+//! module_import   →   "use" ident ( "::" ident) *
+//!
+//! dimension_expr  →   dim_factor
+//! dim_factor      →   dim_power ( (multiply | divide) dim_power ) *
+//! dim_power       →   dim_primary ( power dim_exponent | unicode_exponent ) ?
+//! dim_exponent    →   number | minus dim_exponent | "(" dim_exponent ( divide dim_exponent ) ? ")"
```

Comment:
Numbers in dimension exponents are not as flexible as normal numbers. They are basically just integers.